### FE630 - Homework #1

**Author**: Sid Bhatia

**Date**: March 3rd, 2023

**Pledge**: I pledge my honor that I have abided by the Stevens Honor System.

**Professor**: Papa Momar Ndiaye

#### Problem #1 - Utility Functions

##### 1. Log Utility (15 pts)

a. **Certainty Equivalent and Risk Premium**

An investor whose initial wealth is $\$1000$ is offered an opportunity to play a fair game with 2 possible outcomes: winning $\$350$ with a probability of $1/3$ or losing $\$150$ with a probability of $2/3$. The investor utility function is the natural logarithm of his wealth, $u(W) = \ln(W)$.

- What is the Certainty Equivalent of this risky game?
- What is the exact Risk Premium of this risky game?
- How good is the approximation obtained by using a Taylor series expansion?

First, let's define some terms: 
- **Certainty Equivalent (CE)**: The guaranteed amount of money that an investor considers equivalent to a risky gamble. It's the amount where the investor is indifferent between the certain payment and the risky gamble.
- **Risk Premium (RP)**: The difference between the expected return of the risky gamble and the certainty equivalent. It represents the extra amount an investor needs to be willing to participate in the risky gamble.
- **Utility Function**: In this case, $u(W) = \ln(W)$, where $W$ is wealth. This function represents the investor's satisfaction or utility with different levels of wealth.

Given:
- Initial wealth, $W_0 = \$1000$
- Win amount, $W_{ \text{win} } = \$350$ with probability $p_{\text{win}} = \frac{1}{3}$
- Loss amount, $W_{ \text{loss} } = -\$150$ with probability $p_{\text{loss}} = \frac{2}{3}$

We can calculate the expected utility of the gamble, find the certainty equivalent by equating the expected utility to the utility of the certainty equivalent, and then calculate the risk premium. Additionally, we'll use a Taylor series expansion around the initial wealth to approximate the expected utility and compare it.

As such, we will do the following steps:
1. Calculate the *expected utility* of the game.
2. Find the *certainty equivalent (CE)* by setting the expected utility equal to the utility of the CE and solving for CE.
3. Calculate the *risk premium* as the difference between the expected wealth and the certainty equivalent.
4. Compare the *exact expected utility* to the approximation obtained by using a *Taylor series expansion* around the initial wealth.

In [8]:
import numpy as np
from scipy.optimize import fsolve
from typing import Callable

def utility(W: float) -> float:
    """
    Calculate the utility of wealth using a natural logarithm.

    Parameters:
    - W (float): Wealth.

    Returns:
    - float: Utility of the wealth.
    """
    return np.log(W)

def expected_utility(W0: float, W_win: float, W_lose: float, p_win: float, p_lose: float) -> float:
    """
    Calculate the expected utility of a gamble with given outcomes and probabilities.

    Parameters:
    - W0 (float): Initial wealth.
    - W_win (float): Winning amount.
    - W_lose (float): Losing amount.
    - p_win (float): Probability of winning.
    - p_lose (float): Probability of losing.

    Returns:
    - float: Expected utility of the gamble.
    """
    return p_win * utility(W0 + W_win) + p_lose * utility(W0 + W_lose)

def find_certainty_equivalent(EU: float, W0: float) -> float:
    """
    Find the certainty equivalent of a gamble based on its expected utility.

    Parameters:
    - EU (float): Expected utility of the gamble.
    - W0 (float): Initial wealth.

    Returns:
    - float: Certainty equivalent of the gamble.
    """
    func = lambda CE: utility(W0 + CE) - EU
    CE, = fsolve(func, 0)  # Initial guess for CE is 0.
    return CE

def calculate_risk_premium(E_W: float, W0: float, CE: float) -> float:
    """
    Calculate the risk premium of the gamble.

    Parameters:
    - E_W (float): Expected wealth from the gamble.
    - W0 (float): Initial wealth.
    - CE (float): Certainty equivalent of the gamble.

    Returns:
    - float: Risk premium of the gamble.
    """
    return E_W - (W0 + CE)

def taylor_series_expansion(u_prime: float, u_double_prime: float, delta_W: float) -> float:
    """
    Approximates the expected utility using Taylor series expansion.
    
    Parameters:
    - u_prime: First derivative of the utility function at initial wealth.
    - u_double_prime: Second derivative of the utility function at initial wealth.
    - delta_W: Change in wealth.
    
    Returns:
    - Taylor series approximation of expected utility.
    """
    return u_prime * delta_W + 0.5 * u_double_prime * (delta_W**2)

# Given values
W0 = 1000
W_win = 350
W_lose = -150
p_win = 1/3
p_lose = 2/3

# Calculate expected utility of the gamble.
EU = expected_utility(W0, W_win, W_lose, p_win, p_lose)

# Solve for the certainty equivalent.
CE = find_certainty_equivalent(EU, W0)

# Calculate expected wealth from the game.
E_W = W0 + p_win * W_win + p_lose * W_lose

# Calculate the risk premium.
RP = calculate_risk_premium(E_W, W0, CE)

# Compute the series expansion.
u_prime_W0 = 1/W0
u_double_prime_W0 = -1/(W0**2)
delta_W = p_win * W_win + p_lose * W_lose
EU_approx = utility(W0) + taylor_series_expansion(u_prime_W0, u_double_prime_W0, delta_W)

# Display results.
print(f"Certainty Equivalent: {CE:.2f}")
print(f"Risk Premium: {RP:.2f}")
print(f"Expected Utility: {EU:.4f}")
print(f"Taylor Series Expected Utility Approximation: {EU_approx:.4f}")

Certainty Equivalent: -8.28
Risk Premium: 24.94
Expected Utility: 6.8994
Taylor Series Expected Utility Approximation: 6.9243


1. *What is the Certainty Equivalent of this risky game?*

The Certainty Equivalent (CE) of this risky game is $\bf{-\$8.28}$. This means that the investor would be indifferent between accepting a certain outcome that effectively reduces their wealth to $\$991.72 (\$1000 - \$8.28)$ and taking the gamble with the potential to win $\$350$ or lose $\$150$. The negative value of the CE indicates that, according to the investor's utility function (which is the natural logarithm of wealth in this case), they perceive the gamble as having a net negative effect on their utility compared to their initial position, even though the game is "fair" in terms of expected monetary value. This highlights the risk-averse nature of the utility function, where the potential loss weighs more heavily on the decision than the equivalent potential gain.

2. *What is the exact Risk Premium of this risky game?*

The Risk Premium (RP) for this game is approximately $\bf{\$24.94}$. This is the amount the investor would require as compensation to be indifferent between taking the gamble and not participating, based on their initial wealth of $\$1000$. It represents the monetary value of the risk associated with the gamble to the investor. Since the gamble has a negative certainty equivalent, the risk premium indicates how much the investor values certainty over the possibility of the outcomes provided by the gamble. In simpler terms, it's the price the investor is willing to pay to avoid the risk, reflecting their aversion to the gamble's uncertainty.

3. *How good is the approximation obtained by using a Taylor series expansion?*

The approximation of the expected utility using a Taylor series expansion is **quite good**. The exact expected utility (EU) calculated is **6.8994**, while the Taylor series expected utility approximation (EU_approx) is **6.9243**. The difference between them is minimal, indicating that the Taylor series expansion provides a close approximation to the actual expected utility for this particular setup. This suggests that for small changes in wealth (such as those involved in this gamble, relative to the investor's initial wealth), the Taylor series expansion around the initial wealth effectively captures the utility function's curvature and provides a reliable estimate of the expected utility.

b. **Sensitivity to Initial Wealth**

Assume that the initial wealth is $\$2000$. What is the risk premium now?

In [9]:
W0_new = 2000

# Calculate expected utility of the gamble with new initial wealth.
EU_new = expected_utility(W0_new, W_win, W_lose, p_win, p_lose)

# Solve for the certainty equivalent with new initial wealth.
CE_new = find_certainty_equivalent(EU_new, W0_new)

# Calculate expected wealth from the game with new initial wealth.
E_W_new = W0_new + p_win * W_win + p_lose * W_lose

# Calculate the risk premium with the new initial wealth.
RP_new = calculate_risk_premium(E_W_new, W0_new, CE_new)

print(f"New Risk Premium: {RP_new:.2f}")

New Risk Premium: 13.10


Given the initial wealth increase to $\$2000$ and applying the previously defined Python code, the risk premium for this new scenario is approximately $\bf{\$13.10}$. This result indicates how sensitive the investor's risk aversion is to changes in initial wealth. With an initial wealth of $\$2000$, the investor perceives the gamble to be less risky compared to when the initial wealth was $\$1000$, hence the lower risk premium.

This decrease in the risk premium with an increase in initial wealth can be attributed to the utility function's properties, which in this case is a logarithmic function. The logarithmic utility function exhibits **diminishing marginal utility**, meaning that as wealth increases, the utility gained from an additional dollar becomes less significant. Consequently, the investor's aversion to losing a certain amount of money diminishes as their wealth increases, leading to a lower risk premium for the gamble at higher levels of initial wealth.

c. **Sensitivity to Volatility**

Assume that the initial wealth is $\$1000$ and the outcomes are winning $\$300$ with a probability of $1/2$ or losing $\$300$ with a probability of $1/2$. What is the risk premium now?

In [10]:
# Re-defining necessary values and functions to calculate the new risk premium.

# Updated game conditions for sensitivity to volatility.
W0 = 1000  # Initial wealth is $1000
W_win = 300  # Winning amount is now $300
W_lose = -300  # Losing amount is also $300, indicating increased volatility
p_win = 1/2  # Probability of winning is 1/2
p_lose = 1/2  # Probability of losing is also 1/2

# Calculate expected utility with the new conditions.
EU_new = expected_utility(W0, W_win, W_lose, p_win, p_lose)

# Solve for the certainty equivalent with the new conditions.
CE_new = find_certainty_equivalent(EU_new, W0)

# Calculate expected wealth from the game with the new conditions.
E_W_new = W0 + p_win * W_win + p_lose * W_lose

# Calculate the risk premium with the new conditions
RP_new = calculate_risk_premium(E_W_new, W0, CE_new)

print(f"New Risk Premium: {RP_new:.2f}")

New Risk Premium: 46.06


With the updated game conditions, the risk premium now is approximately $\bf{\$46.06}$. This indicates a significant increase in the risk premium compared to the previous scenarios.

This increase reflects the heightened sensitivity of the investor's risk aversion to the increased volatility of the game. With both potential gains and losses being larger and equally probable, the gamble is perceived as **riskier**, leading to a **higher risk premium**. The investor would require a larger compensation (in terms of a risk premium) to be willing to participate in this more volatile game.

##### 2. Certainty Equivalent and Risk Premium for a Power Utility (30 pts)

An investor whose initial wealth is $\$1000$ is offered an opportunity to play a gamble with 2 possible outcomes: winning $\$250$ with a probability of $2/3$ or losing $\$400$ with a probability of $1/3$. The investor utility function is given by $u(W) = W^k$, where $k \in \mathbb{R}$.

a. **Absolute Risk Aversion & Risk Attitude**

Compute the absolute risk aversion coefficient and describe the risk attitude of the investor (Risk-averse or Risk-taker) as a function of $k$. Justify your answer.

First, let's define **absolute risk aversion** (**ARA**) and explain how it relates to the utility function's shape as well as the parameter $k$.

$$\text{ARA}(W) = -\frac{u''(W)}{u'(W)}$$

Given the utility function $u(W) = W^k$, we can compute its *first* and *second derivatives* to find the **absolute risk aversion coefficient**:

- First derivative $u'(W) = k \cdot W^{k - 1}$
- Second derivative $u''(W) = k \cdot (k - 1) \cdot W^{k - 2}$

Substituting these into the formula for absolute risk aversion, we get:

$$\text{ARA}(W) = -\frac{k \cdot (k - 1) \cdot W^{k - 2}}{k \cdot W^{k - 1}} = -(k - 1) \cdot W^{-1} = -\frac{k - 1}{W} = \boldsymbol{\frac{1 - k}{W}}$$

The sign and magnitude of the ARA coefficient depend on $k$ and $W$, which indicates the investor's risk attitude:

- If $k < 1 , \text{ARA}(W) > 0$, indicating *risk aversion* since the utility function is **concave**.
- If $k > 1, \text{ARA}(W) < 0$, indicating *risk seeking* since the utility function is **convex**.
- If $k = 1, \text{ARA}(W) = 0$, indicating *risk neutrality* since the utility function is **linear**.

In [12]:
def compute_ARA_and_risk_attitude(k: float, W: float = 1000) -> None:
    """
    Computes the Absolute Risk Aversion (ARA) coefficient for a given k and W,
    and describes the investor's risk attitude.
    
    Parameters:
    - k: The exponent in the utility function u(W) = W^k.
    - W: Wealth, with a default value of 1000.
    """
    # First derivative of the utility function
    u_prime = k * W**(k - 1)
    
    # Second derivative of the utility function
    u_double_prime = k * (k - 1) * W**(k - 2)
    
    # Absolute Risk Aversion (ARA) coefficient
    ARA = -(u_double_prime / u_prime)
    
    # Determine risk attitude based on k
    if k < 1:
        risk_attitude = "Risk-averse"
    elif k > 1:
        risk_attitude = "Risk-taker"
    else:  # k == 1
        risk_attitude = "Risk-neutral"
    
    print(f"For k = {k}, the Absolute Risk Aversion (ARA) coefficient is {ARA:.4f}, indicating the investor is {risk_attitude}.")

# Example usage
k_values = [0.5, 1, 1.5]  # Example k values to illustrate different risk attitudes
for k in k_values:
    compute_ARA_and_risk_attitude(k)

For k = 0.5, the Absolute Risk Aversion (ARA) coefficient is 0.0005, indicating the investor is Risk-averse.
For k = 1, the Absolute Risk Aversion (ARA) coefficient is -0.0000, indicating the investor is Risk-neutral.
For k = 1.5, the Absolute Risk Aversion (ARA) coefficient is -0.0005, indicating the investor is Risk-taker.


1. **Risk-averse**:

When $k < 1$, the utility function $u(W) = W^k$ is **concave**. In this case, the investor exhibits **diminishing marginal utility of wealth**, meaning that the utility gained from each additional unit of wealth decreases as wealth increases. This concavity leads to a positive ARA coefficient, as the investor values a sure outcome more than an uncertain one with the same expected value. The positive ARA coefficient $(\text{ARA} > 0)$ for $k < 1$ indicates that the investor is risk-averse, preferring certainty to risk when faced with potential gains or losses.

2. **Risk-neutral**:

When $k = 1$, the utility function $u(W) = W^k = W$ is **linear**. This linearity implies **constant marginal utility of wealth**, where the investor values each additional unit of wealth equally, regardless of their current wealth level. A linear utility function results in an ARA coefficient of zero $(\text{ARA} = 0)$, indicating that the investor is risk-neutral. A risk-neutral investor is indifferent between a certain outcome and a gamble with the same expected return, showing no preference for or aversion to risk.

3. **Risk-taker (Risk-seeking)**:

When $k > 1$, the utility function $u(W) = W^k$ is **convex**. This convexity reflects **increasing marginal utility of wealth**, meaning that the utility gained from each additional unit of wealth increases as wealth increases. Such a utility function leads to a negative ARA coefficient $(\text{ARA} < 0)$, indicating that the investor is a risk-taker or risk-seeking. Risk-takers derive more utility from the possibility of higher gains associated with risk than from the certainty of a sure outcome, even when both options have the same expected return.